vkubelet/podcontroller.go

```diff
@@ -0,0 +1,380 @@
+		span.SetStatus(ocstatus.FromError(err))
+		log.G(ctx).Error(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to converge the two.
+func (pc *PodController) syncHandler(ctx context.Context, key string) error {
+	// Create a span for the current call.
+	ctx, span := trace.StartSpan(ctx, "syncHandler")
+	defer span.End()
+
+	// Add the current key as an attribute to the current span.
+	span.AddAttributes(trace.StringAttribute("key", key))
+
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// Log the error as a warning, but do not requeue the key as it is invalid.
+		log.G(ctx).Warn(pkgerrors.Wrapf(err, "invalid resource key: %q", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name.
+	pod, err := pc.podsLister.Pods(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// We've failed to fetch the pod from the lister, but the error is not a 404.
+			// Hence, we add the key back to the work queue so we can retry processing it later.
+			err := pkgerrors.Wrapf(err, "failed to fetch pod with key %q from lister", key)
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		// At this point we know the Pod resource doesn't exist, which most probably means it was deleted.
+		// Hence, we must delete it from the provider if it still exists there.
+		if err := pc.server.deletePod(ctx, namespace, name); err != nil {
+			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodNameFromCoordinates(namespace, name))
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		return nil
+	}
+	// At this point we know the Pod resource has either been created or updated (which includes being marked for deletion).
+	return pc.syncPodInProvider(ctx, pod)
+}
+
+// syncPodInProvider tries and reconciles the state of a pod by comparing its Kubernetes representation and the provider's representation.
+func (pc *PodController) syncPodInProvider(ctx context.Context, pod *corev1.Pod) error {
+	// Create a span for the current call.
+	ctx, span := trace.StartSpan(ctx, "syncPodInProvider")
+	defer span.End()
+
+	// Add the pod's attributes to the current span.
+	addPodAttributes(span, pod)
+
+	// Check whether the pod has been marked for deletion.
+	// If it does, delete it in the provider.
+	if pod.DeletionTimestamp != nil {
+		// Delete the pod.
+		if err := pc.server.deletePod(ctx, pod.Namespace, pod.Name); err != nil {
+			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodName(pod))
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		return nil
+	}
+
+	// Ignore the pod if it is in the "Failed" state.
+	if pod.Status.Phase == corev1.PodFailed {
+		log.G(ctx).Warnf("skipping sync of pod %q in %q phase", loggablePodName(pod), pod.Status.Phase)
```

Comment:
Should this be debug?

Reply:
IMHO a pod in `Failed` state is enough reason to issue a warning to the user, especially in a situation where we're ignoring an "event". If we change this to debug, the user will most probably never be aware of this fact.